execution_engine/src/core/runtime/externals.rs

```diff
@@ -1097,6 +1098,43 @@ where
 
                 Ok(Some(RuntimeValue::I32(api_error::i32_from(result))))
             }
+
+            FunctionIndex::GenericHash => {
+                // args(0) = pointer to input in Wasm memory
+                // args(1) = size of input in Wasm memory
+                // args(2) = integer representation of HashAlgoType enum variant
+                // args(3) = pointer to output pointer in Wasm memory
+                // args(4) = size of output
+                let (in_ptr, in_size, hash_algo_type, out_ptr, out_size) = Args::parse(args)?;
+                self.charge_host_function_call(
+                    &host_function_costs.generic_hash,
+                    [in_ptr, in_size, hash_algo_type, out_ptr, out_size],
+                )?;
+                let hash_algo_type =
+                    HashAlgoType::try_from(hash_algo_type as u8).map_err(|e| Error::from(e))?;
+
+                let digest =
+                    self.checked_memory_slice(in_ptr as usize, in_size as usize, |input| {
+                        match hash_algo_type {
+                            HashAlgoType::Blake2b => crypto::blake2b(input),
+                            HashAlgoType::Blake3 => core_crypto::blake3(input),
+                        }
+                    })?;
+
+                let result = if digest.len() > out_size as usize {
```

Comment:
Something I just noticed - we've done hashing already, but we're checking the size after the fact. Since the output of the hash is constant, we could also move this check above the `checked_memory_slice` line.

Reply:
Even though the output is always expected to be 32 bytes, those hashes use different constants for their digest size:
- blake2: `BLAKE2B_DIGEST_LENGTH`
- blake3: `BLAKE3_DIGEST_LENGTH`

Unless we implement all the necessary hash functions (e.g., _sha256_, _keccak_, _poseidon_) and **generalize** it with something like `GENERIC_HASH_DIGEST_LENGTH`, I would prefer to avoid using hardcoded pre-checks.